File: frontend/src/app/services/tmdb.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, catchError, throwError } from 'rxjs';
import { environment } from '../../environments/environment';
import {
  Movie,
  TVShow,
  TMDBResponse,
  MovieDetails,
  TVShowDetails,
  Genre,
  Credits,
  ReviewsResponse,
  VideosResponse,
} from '../models/tmdb.model';

@Injectable({
  providedIn: 'root',
})
export class TmdbService {
  private readonly BASE_URL = environment.tmdbBaseUrl;
  private readonly API_KEY = environment.tmdbApiKey;
  private readonly IMAGE_BASE_URL = environment.tmdbImageBaseUrl;

  constructor(private http: HttpClient) {}

  private getParams(additionalParams: { [key: string]: string } = {}): HttpParams {
    let params = new HttpParams().set('api_key', this.API_KEY);
    Object.keys(additionalParams).forEach((key) => {
      params = params.set(key, additionalParams[key]);
    });
    return params;
  }

  getPosterUrl(path: string | null, size: string = 'w500'): string {
    return path ? `${this.IMAGE_BASE_URL}/${size}${path}` : '/assets/no-image.png';
  }

  getBackdropUrl(path: string | null, size: string = 'original'): string {
    return path ? `${this.IMAGE_BASE_URL}/${size}${path}` : '/assets/no-image.png';
  }

  getTrendingMovies(timeWindow: 'day' | 'week' = 'week'): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/trending/movie/${timeWindow}`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getPopularMovies(page: number = 1): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/movie/popular`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getTopRatedMovies(page: number = 1): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/movie/top_rated`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getUpcomingMovies(page: number = 1): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/movie/upcoming`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getMovieDetails(movieId: number): Observable<MovieDetails> {
    return this.http
      .get<MovieDetails>(`${this.BASE_URL}/movie/${movieId}`, { params: this.getParams() })
      .pipe(catchError(this.handleError));
  }

  searchMovies(query: string, page: number = 1): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/search/movie`, {
        params: this.getParams({ query, page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  searchMulti(query: string, page: number = 1): Observable<TMDBResponse<any>> {
    return this.http
      .get<TMDBResponse<any>>(`${this.BASE_URL}/search/multi`, {
        params: this.getParams({ query, page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getTrendingTVShows(timeWindow: 'day' | 'week' = 'week'): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/trending/tv/${timeWindow}`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getPopularTVShows(page: number = 1): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/tv/popular`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getTopRatedTVShows(page: number = 1): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/tv/top_rated`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getTVShowDetails(tvId: number): Observable<TVShowDetails> {
    return this.http
      .get<TVShowDetails>(`${this.BASE_URL}/tv/${tvId}`, { params: this.getParams() })
      .pipe(catchError(this.handleError));
  }

  searchTVShows(query: string, page: number = 1): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/search/tv`, {
        params: this.getParams({ query, page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getMovieGenres(): Observable<{ genres: Genre[] }> {
    return this.http
      .get<{ genres: Genre[] }>(`${this.BASE_URL}/genre/movie/list`, { params: this.getParams() })
      .pipe(catchError(this.handleError));
  }

  getTVGenres(): Observable<{ genres: Genre[] }> {
    return this.http
      .get<{ genres: Genre[] }>(`${this.BASE_URL}/genre/tv/list`, { params: this.getParams() })
      .pipe(catchError(this.handleError));
  }

  discoverMovies(params: { [key: string]: string } = {}): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/discover/movie`, {
        params: this.getParams(params),
      })
      .pipe(catchError(this.handleError));
  }

  discoverTVShows(params: { [key: string]: string } = {}): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/discover/tv`, { params: this.getParams(params) })
      .pipe(catchError(this.handleError));
  }

  getMovieCredits(movieId: number): Observable<Credits> {
    return this.http
      .get<Credits>(`${this.BASE_URL}/movie/${movieId}/credits`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getTVShowCredits(tvId: number): Observable<Credits> {
    return this.http
      .get<Credits>(`${this.BASE_URL}/tv/${tvId}/credits`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getMovieReviews(movieId: number, page: number = 1): Observable<ReviewsResponse> {
    return this.http
      .get<ReviewsResponse>(`${this.BASE_URL}/movie/${movieId}/reviews`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getTVShowReviews(tvId: number, page: number = 1): Observable<ReviewsResponse> {
    return this.http
      .get<ReviewsResponse>(`${this.BASE_URL}/tv/${tvId}/reviews`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getSimilarMovies(movieId: number, page: number = 1): Observable<TMDBResponse<Movie>> {
    return this.http
      .get<TMDBResponse<Movie>>(`${this.BASE_URL}/movie/${movieId}/similar`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getSimilarTVShows(tvId: number, page: number = 1): Observable<TMDBResponse<TVShow>> {
    return this.http
      .get<TMDBResponse<TVShow>>(`${this.BASE_URL}/tv/${tvId}/similar`, {
        params: this.getParams({ page: page.toString() }),
      })
      .pipe(catchError(this.handleError));
  }

  getMovieVideos(movieId: number): Observable<VideosResponse> {
    return this.http
      .get<VideosResponse>(`${this.BASE_URL}/movie/${movieId}/videos`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getTVShowVideos(tvId: number): Observable<VideosResponse> {
    return this.http
      .get<VideosResponse>(`${this.BASE_URL}/tv/${tvId}/videos`, {
        params: this.getParams(),
      })
      .pipe(catchError(this.handleError));
  }

  getProfileUrl(path: string | null, size: string = 'w185'): string {
    return path ? `${this.IMAGE_BASE_URL}/${size}${path}` : '/assets/no-avatar.png';
  }

  private handleError(error: any): Observable<never> {
    console.error('TMDB API Error:', error);
    return throwError(() => error);
  }
}

File: flutter_app/lib/services/tmdb_service.dart
// filepath: lib/services/tmdb_service.dart
import 'package:dio/dio.dart';
import '../models/tmdb_models.dart';
import '../config/app_config.dart';

class TmdbService {
  final Dio _dio;

  TmdbService(this._dio);

  Future<MovieDetails> getMovieDetails(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/movie/$id', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      return MovieDetails.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to load movie details: $e');
    }
  }

  Future<List<Cast>> getMovieCredits(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/movie/$id/credits', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['cast'] as List;
      return list.map((e) => Cast.fromJson(e)).toList();
    } catch (e) {
      throw Exception('Failed to load movie credits: $e');
    }
  }

  Future<List<Video>> getMovieVideos(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/movie/$id/videos', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['results'] as List;
      return list
          .map((e) => Video.fromJson(e))
          .where((v) => v.site == 'YouTube')
          .toList();
    } catch (e) {
      throw Exception('Failed to load movie videos: $e');
    }
  }

  Future<List<MovieDetails>> getSimilarMovies(int id) async {
     try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/movie/$id/similar', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['results'] as List;
      return list.map((e) => MovieDetails.fromJson(e)).toList();
    } catch (e) {
      throw Exception('Failed to load similar movies: $e');
    }
  }

  Future<List<Review>> getMovieReviews(int id) async {
     try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/movie/$id/reviews', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['results'] as List;
      return list.map((e) => Review.fromJson(e)).toList();
    } catch (e) {
      throw Exception('Failed to load movie reviews: $e');
    }
  }

  Future<TVShowDetails> getTVShowDetails(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/tv/$id', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      return TVShowDetails.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to load TV show details: $e');
    }
  }

  Future<List<Cast>> getTVShowCredits(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/tv/$id/credits', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['cast'] as List;
      return list.map((e) => Cast.fromJson(e)).toList();
    } catch (e) {
      throw Exception('Failed to load TV show credits: $e');
    }
  }

   Future<List<Video>> getTVShowVideos(int id) async {
    try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/tv/$id/videos', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['results'] as List;
      return list
          .map((e) => Video.fromJson(e))
          .where((v) => v.site == 'YouTube')
          .toList();
    } catch (e) {
      throw Exception('Failed to load TV show videos: $e');
    }
  }

  Future<List<TVShowDetails>> getSimilarTVShows(int id) async {
     try {
      final response = await _dio.get('${AppConfig.tmdbBaseUrl}/tv/$id/similar', queryParameters: {
        'api_key': AppConfig.tmdbApiKey,
      });
      final list = response.data['results'] as List;
      return list.map((e) => TVShowDetails.fromJson(e)).toList();
    } catch (e) {
      throw Exception('Failed to load similar TV shows: $e');
    }
  }
}


